app/src/main/java/com/strv/movies/ui/moviedetail/MovieDetailViewModel.kt

```diff
@@ -5,30 +5,39 @@ import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import com.strv.movies.extension.fold
+import com.strv.movies.model.MovieDetail
 import com.strv.movies.network.MovieRepository
 import com.strv.movies.ui.navigation.MoviesNavArguments
 import dagger.hilt.android.lifecycle.HiltViewModel
 import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.combine
 import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class MovieDetailViewModel @Inject constructor(
     savedStateHandle: SavedStateHandle,
-    movieRepository: MovieRepository
+    private val movieRepository: MovieRepository
 ) : ViewModel() {
 
     private val movieId =
         requireNotNull(savedStateHandle.get<Int>(MoviesNavArguments.MOVIE_ID_KEY)) {
             "Movie id is missing..."
         }
 
+    private val _movieDetail = MutableStateFlow<MovieDetail?>(null)
     private val _viewState = MutableStateFlow(MovieDetailViewState(loading = true))
-    val viewState = _viewState.asStateFlow()
+    val viewState = combine(_viewState, _movieDetail) { state, detail ->
+        when {
+            detail != null -> MovieDetailViewState(movie = detail)
+            else -> state
+        }
+    }
 
     init {
+        observeMovieDetail()
+
         viewModelScope.launch {
             movieRepository.getMovieDetail(movieId).fold(
```

Comment:
You should only observe movie detail in ViewModel. Combining sources from the database and network should be in the repository.

Reply:
That's exactly what I do! But you have to consider several cases:
 - when you open detail for the first time -> you have no data
 - if you have no data -> progress should be shown
 - if you have no data & API fails -> show error

Otherwise, I would have to show an empty screen no matter what is happing... So what I do is:
If you have some data, show it immediately. Else reflect the state from network call.